src/util/codec/mysql/json.rs

```diff
@@ -0,0 +1,545 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use super::Result;
+use std::{str, f64};
+use std::collections::BTreeMap;
+use std::result::Result as SResult;
+use serde_json::{self, Value};
+use std::io::Read;
+use util::codec::number::{NumberDecoder, NumberEncoder};
+use serde::ser::{Serialize, Serializer, SerializeTuple, SerializeMap};
+
+// The binary jSON format from MySQL 5.7 is as follows:
+//
+// JSON doc ::= type value
+// type ::=
+// 0x01 |       // large JSON object
+// 0x03 |       // large JSON array
+// 0x04 |       // literal (true/false/null)
+// 0x05 |       // int16
+// 0x06 |       // uint16
+// 0x07 |       // int32
+// 0x08 |       // uint32
+// 0x09 |       // int64
+// 0x0a |       // uint64
+// 0x0b |       // double
+// 0x0c |       // utf8mb4 string
+//
+// value ::=
+// object  |
+// array   |
+// literal |
+// number  |
+// string  |
+//
+// object ::= element-count size key-entry* value-entry* key* value*
+//
+// array ::= element-count size value-entry* value*
+//
+// number of members in object or number of elements in array
+// element-count ::= uint32
+//
+// number of bytes in the binary representation of the object or array
+// size ::= uint32
+//
+// key-entry ::= key-offset key-length
+//
+// key-offset ::= uint32
+//
+// key-length ::= uint16    // key length must be less than 64KB
+//
+// value-entry ::= type offset-or-inlined-value
+//
+// This field holds either the offset to where the value is stored,
+// or the value itself if it is small enough to be inlined (that is,
+// if it is a JSON literal or a small enough [u]int).
+// offset-or-inlined-value ::= uint32
+//
+// key ::= utf8mb4-data
+//
+// literal ::=
+// 0x00 |   // JSON null literal
+// 0x01 |   // JSON true literal
+// 0x02 |   // JSON false literal
+//
+// number ::=  ....    // little-endian format for [u]int(16|32|64), whereas
+// double is stored in a platform-independent, eight-byte
+// format using float8store()
+//
+// string ::= data-length utf8mb4-data
+//
+// data-length ::= uint8*    // If the high bit of a byte is 1, the length
+// field is continued in the next byte,
+// otherwise it is the last byte of the length
+// field. So we need 1 byte to represent
+// lengths up to 127, 2 bytes to represent
+// lengths up to 16383, and so on...
+//
+
+
+const TYPE_CODE_OBJECT: u8 = 0x01;
+const TYPE_CODE_ARRAY: u8 = 0x03;
+const TYPE_CODE_LITERAL: u8 = 0x04;
+const TYPE_CODE_I64: u8 = 0x09;
+const TYPE_CODE_DOUBLE: u8 = 0x0b;
+const TYPE_CODE_STRING: u8 = 0x0c;
+
+const JSON_LITERAL_NIL: u8 = 0x00;
+const JSON_LITERAL_TRUE: u8 = 0x01;
+const JSON_LITERAL_FALSE: u8 = 0x02;
+
+const LENGTH_TYPE: usize = 1;
+const LENGTH_LITERAL: usize = 1;
+const LENGTH_U16: usize = 2;
+const LENGTH_U32: usize = 4;
+const LENGTH_KEY_ENTRY: usize = LENGTH_U32 + LENGTH_U16;
+const LENGTH_VALUE_ENTRY: usize = LENGTH_TYPE + LENGTH_U32;
+
+macro_rules! take_prefix {
```

Comment:
I don't think the macro is necessary. In most case, use `&data[..len]` is enough.